agent/proxycfg/state.go

```diff
@@ -843,6 +886,223 @@ func (s *state) resetWatchesFromChain(
+
+				if err != nil {
+					logger.Error("failed to register watch for a service-leaf",
+						"service", svc.Service.String(),
+						"error", err,
+					)
+					cancel()
+					return err
+				}
+				snap.TerminatingGateway.WatchedLeaves[svc.Service] = cancel
+			}
+
+			// Watch service resolvers for the service
+			// These are used to create clusters and endpoints for the service subsets
+			if _, ok := snap.TerminatingGateway.WatchedResolvers[svc.Service]; !ok {
+				ctx, cancel := context.WithCancel(s.ctx)
+				err := s.cache.Notify(ctx, cachetype.ConfigEntriesName, &structs.ConfigEntryQuery{
+					Datacenter:     s.source.Datacenter,
+					QueryOptions:   structs.QueryOptions{Token: s.token},
+					Kind:           structs.ServiceResolver,
+					Name:           svc.Service.ID,
+					EnterpriseMeta: svc.Service.EnterpriseMeta,
+				}, fmt.Sprintf("service-resolver:%s", svc.Service.String()), s.ch)
+
+				if err != nil {
+					logger.Error("failed to register watch for a service-resolver",
+						"service", svc.Service.String(),
+						"error", err,
+					)
+					cancel()
+					return err
+				}
+				snap.TerminatingGateway.WatchedResolvers[svc.Service] = cancel
+			}
+		}
+
+		// Cancel service instance watches for services that were not in the update
+		for sid, cancelFn := range snap.TerminatingGateway.WatchedServices {
+			if _, ok := svcMap[sid]; !ok {
+				logger.Debug("canceling watch for service", "service", sid.String())
+				delete(snap.TerminatingGateway.WatchedServices, sid)
+				delete(snap.TerminatingGateway.ServiceGroups, sid)
+				cancelFn()
+			}
+		}
+
+		// Cancel leaf cert watches for services that were not in the update
+		for sid, cancelFn := range snap.TerminatingGateway.WatchedLeaves {
+			if _, ok := svcMap[sid]; !ok {
+				logger.Debug("canceling watch for leaf cert", "service", sid.String())
+				delete(snap.TerminatingGateway.WatchedLeaves, sid)
+				delete(snap.TerminatingGateway.ServiceLeaves, sid)
+				cancelFn()
+			}
+		}
+
+		// Cancel service-resolver watches for services that were not in the update
+		for sid, cancelFn := range snap.TerminatingGateway.WatchedResolvers {
+			if _, ok := svcMap[sid]; !ok {
+				logger.Debug("canceling watch for service-resolver", "service", sid.String())
+				delete(snap.TerminatingGateway.WatchedResolvers, sid)
+				delete(snap.TerminatingGateway.ServiceResolvers, sid)
+				cancelFn()
+			}
+		}
+
+		// Cancel intention watches for services that were not in the update
+		for sid, cancelFn := range snap.TerminatingGateway.WatchedIntentions {
+			if _, ok := svcMap[sid]; !ok {
+				logger.Debug("canceling watch for intention", "service", sid.String())
+				delete(snap.TerminatingGateway.WatchedIntentions, sid)
+
+				// No additional deletions needed, since intentions aren't stored in snapshot
+
+				cancelFn()
+			}
+		}
+
+	default:
+		switch {
+		// Store service instances for watched service
+		case strings.HasPrefix(u.CorrelationID, "external-service:"):
```

Comment:
nit: I feel like just having every case at the top-level switch here is more readable, that way we don't have to indent as much.

i.e.
```
switch switch u.CorrelationID {
case rootsWatchID:
...
case strings.HasPrefix(u.CorrelationID, "external-service:"):
}
```

instead of having a second switch statement here.

Reply:
I think the top level switch won't be able to use the variable, but that's ok. I generally will take this approach as well:

```go
switch {
case u.CorrelationID == rootsWatchID:
...
case strings.HasPrefix(u.CorrelationID, "external-service:"):
}
```

Reply:
Sure thing, just updated it.